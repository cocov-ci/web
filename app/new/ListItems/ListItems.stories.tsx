import type { Meta, StoryObj } from '@storybook/react'

import ListItems from './index'

const meta: Meta<typeof ListItems> = {
  title: 'Repositories/New/ListItems',
  component: ListItems,
  tags: ['autodocs'],
}

export default meta

type Story = StoryObj<typeof ListItems>

export const Default: Story = {
  args: {
    data: [
      {
        name: '.github',
        description: 'Community health files for the @cocov-ci organization',
        created_at: '2023-01-03T19:43:48.000Z',
        pushed_at: '2023-01-03T20:12:13.000Z',
        status: 'absent',
      },
      {
        name: 'api',
        description: "Cocov's API.",
        created_at: '2022-11-22T21:06:17.000Z',
        pushed_at: '2023-03-06T01:09:42.000Z',
        status: 'present',
      },
      {
        name: 'badger',
        description: "Cocov's Badge Generator Service",
        created_at: '2023-01-03T19:20:17.000Z',
        pushed_at: '2023-01-03T19:20:33.000Z',
        status: 'present',
      },
      {
        name: 'brakeman',
        description: 'Brakeman Plugin for Cocov',
        created_at: '2022-12-14T22:35:46.000Z',
        pushed_at: '2023-01-07T21:33:21.000Z',
        status: 'absent',
      },
    ],
  },
}
