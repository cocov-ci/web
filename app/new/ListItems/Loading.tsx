import classNames from 'classnames'

import Loading from 'app/common/Loading'
import randomBetween from 'utils/randomBetween'

import styles from './ListItems.module.scss'

const LoadingComponent = () => {
  return (
    <div>
      {Array.from(Array(7).keys()).map(item => (
        <div className={classNames(styles.item, styles.loading)} key={item}>
          <Loading
            className={styles.icon}
            height="27px"
            type="skeleton"
            width="27px"
          />
          <div className={styles.data}>
            <Loading
              alignment="left"
              height="32px"
              type="skeleton"
              width={`${randomBetween(100, 250)}px`}
            />
            <Loading
              alignment="left"
              height="20px"
              type="skeleton"
              width={`${randomBetween(250, 350)}px`}
            />
            <Loading
              alignment="left"
              className={styles.timestamp}
              height="20px"
              type="skeleton"
              width={`${randomBetween(250, 350)}px`}
            />
          </div>
          <Loading
            className={styles.button}
            height="32px"
            type="skeleton"
            width="150px"
          />
        </div>
      ))}
    </div>
  )
}

export default LoadingComponent
