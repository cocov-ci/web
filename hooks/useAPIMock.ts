import API from '../utils/api'

const useAPIMock = <I, O>(fn: (params: I) => Promise<O>, result: O) => {
  jest
    .spyOn(API.shared, fn.name as keyof API)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    .mockImplementation(() => Promise.resolve(result) as any)
}

export default useAPIMock
