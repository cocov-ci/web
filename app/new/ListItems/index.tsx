'use client'

import { Album } from 'lucide-react'
import { useState } from 'react'

import Button from 'app/common/Button'
import RelativeTime from 'app/common/RelativeTime'
import Text from 'app/common/Text'
import { useErrorBanner } from 'hooks/useBanner'
import { OrgRepositoryProps } from 'types/Repositories'
import API from 'utils/api'

import styles from './ListItems.module.scss'
import Loading from './Loading'

interface ItemParams {
  item: OrgRepositoryProps
  onAddSuccess: () => void
}

interface ListItemsParams {
  data: OrgRepositoryProps[]
  loading: boolean
  refetch: () => void
}

const Item = ({ item, onAddSuccess }: ItemParams) => {
  const { showBanner } = useErrorBanner()
  const [addingRepository, setAddingRepository] = useState<boolean>(false)

  const onAddRepository = async () => {
    setAddingRepository(true)

    try {
      await API.shared.repositoryAdd({ name: item.name })
      onAddSuccess()
    } catch (err) {
      showBanner({
        children: `We cannot add the repository "${item.name}". Please try again!`,
      })
    } finally {
      setAddingRepository(false)
    }
  }

  return (
    <div className={styles.item}>
      <Album className={styles.icon} size={27} />
      <div className={styles.data}>
        <Text className={styles.name} variant="title">
          {item.name}
        </Text>
        <Text>{item.description}</Text>
        <Text className={styles.timestamp}>
          Created <RelativeTime timestamp={new Date(item.created_at)} />
          {item?.pushed_at && (
            <>
              {' '}
              â€¢ Last updated{' '}
              <RelativeTime timestamp={new Date(item.pushed_at)} />
            </>
          )}
        </Text>
      </div>
      {!addingRepository && item.status === 'absent' && (
        <Button onClick={() => onAddRepository()} style="secondary">
          Add to Cocov
        </Button>
      )}
      {!addingRepository && item.status === 'present' && (
        <Text className={styles.alreadyAddedStatus}>Already added</Text>
      )}
      {addingRepository && (
        <Button disabled style="secondary">
          Adding
        </Button>
      )}
    </div>
  )
}

const ListItems = ({ data, refetch, loading }: ListItemsParams) => {
  if (loading) return <Loading />

  return (
    <div>
      {data?.map(item => (
        <Item item={item} key={item.name} onAddSuccess={() => refetch()} />
      ))}
    </div>
  )
}

export default ListItems
