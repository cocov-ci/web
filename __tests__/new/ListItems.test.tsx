import { fireEvent, render, screen } from '@testing-library/react'
import { act } from 'react-dom/test-utils'

import ListItems from 'app/new/ListItems'
import useAPIMock from 'hooks/useAPIMock'
import API from 'utils/api'

import '@testing-library/jest-dom'

const args = {
  data: [
    {
      name: '.github',
      description: 'Community health files for the @cocov-ci organization',
      created_at: '2023-01-03T19:43:48.000Z',
      pushed_at: '2023-01-03T20:12:13.000Z',
      status: 'absent' as const,
    },
    {
      name: 'api',
      description: "Cocov's API.",
      created_at: '2022-11-22T21:06:17.000Z',
      pushed_at: '2023-03-06T01:09:42.000Z',
      status: 'present' as const,
    },
  ],
  refetch: () => null,
  loading: false,
}

describe('new/ListItems', () => {
  it('renders ListItems', () => {
    const { container } = render(<ListItems {...args} />)

    expect(container.querySelectorAll('.item').length).toEqual(2)
  })

  it('renders ListItems with "loading" param', () => {
    const { container } = render(<ListItems {...args} loading={true} />)

    expect(container.querySelectorAll('.item.loading').length).toEqual(7)
  })

  it('triggers "Add to Cocov" button', async () => {
    const handleEvent = jest.fn()

    render(<ListItems {...args} refetch={handleEvent} />)

    useAPIMock(API.prototype.repositoryAdd, {
      id: 123,
      description: 'test',
      name: 'api',
      token: '1234',
      coverage: 1,
      issues: 3,
      default_branch: 'main',
      head: { checks_status: 'ok', coverage_status: 'ok' },
    })

    await act(async () => {
      fireEvent.click(screen.getByText('Add to Cocov'))
    })

    expect(handleEvent).toHaveBeenCalledTimes(1)
  })

  it('renders ListItems snapshots', () => {
    const { container } = render(<ListItems {...args} />)

    expect(container).toMatchSnapshot()
  })
})
