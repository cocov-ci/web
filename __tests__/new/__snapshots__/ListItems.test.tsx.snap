// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new/ListItems renders ListItems snapshots 1`] = `
<div>
  <div>
    <div
      class="item"
    >
      <svg
        class="icon"
        fill="none"
        height="27"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="27"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          height="18"
          rx="2"
          ry="2"
          width="18"
          x="3"
          y="3"
        />
        <polyline
          points="11 3 11 11 14 8 17 11 17 3"
        />
      </svg>
      <div
        class="data"
      >
        <p
          class="text name title"
        >
          .github
        </p>
        <p
          class="text"
        >
          Community health files for the @cocov-ci organization
        </p>
        <p
          class="text timestamp"
        >
          Created 
          <span
            class=""
            title="Jan 3, 2023, 7:43 PM GMT"
          >
            3 months ago
          </span>
           
          • Last updated
           
          <span
            class=""
            title="Jan 3, 2023, 8:12 PM GMT"
          >
            3 months ago
          </span>
        </p>
      </div>
      <button
        class="className button secondary"
      >
        Add to Cocov
      </button>
    </div>
    <div
      class="item"
    >
      <svg
        class="icon"
        fill="none"
        height="27"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="27"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          height="18"
          rx="2"
          ry="2"
          width="18"
          x="3"
          y="3"
        />
        <polyline
          points="11 3 11 11 14 8 17 11 17 3"
        />
      </svg>
      <div
        class="data"
      >
        <p
          class="text name title"
        >
          api
        </p>
        <p
          class="text"
        >
          Cocov's API.
        </p>
        <p
          class="text timestamp"
        >
          Created 
          <span
            class=""
            title="Nov 22, 2022, 9:06 PM GMT"
          >
            4 months ago
          </span>
           
          • Last updated
           
          <span
            class=""
            title="Mar 6, 2023, 1:09 AM GMT"
          >
            4 weeks ago
          </span>
        </p>
      </div>
      <p
        class="text alreadyAddedStatus"
      >
        Already added
      </p>
    </div>
  </div>
</div>
`;
